src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -268,65 +265,108 @@ class Checker(globalTable: SymbolTable, verbose: Boolean = false) {
+                }
+            case (jvcon1: JavaFFIContractImpl, jvcon2: JavaFFIContractImpl) => jvcon1 == jvcon2
             case (obs: ObsidianContractImpl, jvcon: JavaFFIContractImpl) => false
-            case (jvcon: JavaFFIContractImpl, obs:ObsidianContractImpl) => (obs.name == jvcon.interface)
+            case (jvcon: JavaFFIContractImpl, obs:ObsidianContractImpl) => obs.name == jvcon.interface
         }
     }
 
+    private def typeBound(table: ContractTable): ObsidianType => ObsidianType = {
+        case np: NonPrimitiveType => np match {
+            case GenericType(gVar, bound) => bound.referenceType
+            case t => t
+        }
+
+        case t => t
+    }
 
     /* true iff [t1 <: t2] */
-    private def isSubtype(t1: ObsidianType, t2: ObsidianType, isThis: Boolean): Option[Error] = {
-        (t1, t2) match {
-            case (BottomType(), _) => None
-            case (_, BottomType()) => None
-            case (IntType(), IntType()) => None
-            case (BoolType(), BoolType()) => None
-            case (StringType(), StringType()) => None
+    private def isSubtype(table: ContractTable, t1: ObsidianType, t2: ObsidianType, isThis: Boolean): Option[Error] = {
+        val isSubtypeRes = (typeBound(table)(t1), typeBound(table)(t2)) match {
+            case (BottomType(), _) => true
+            case (_, BottomType()) => true
+            case (IntType(), IntType()) => true
+            case (BoolType(), BoolType()) => true
+            case (StringType(), StringType()) => true
             case (np1: NonPrimitiveType, np2: NonPrimitiveType) =>
-                val mainSubtype: Boolean = (np1, np2) match {
+                (np1, np2) match {
                     case (ContractReferenceType(c1, c1p, _), ContractReferenceType(c2, c2p, _)) =>
-                        contractIsSubtype(c1.contractName, c2.contractName) && isSubpermission(c1p, c2p)
+                        contractIsSubtype(table, c1, c2, isThis) && isSubpermission(c1p, c2p)
                     case (StateType(c1, ss1, _), StateType(c2, ss2, _)) =>
-                        contractIsSubtype(c1, c2) && ss1.subsetOf(ss2)
-                    case (StateType(c, ss1, _), ContractReferenceType(c2, c2p, _)) =>
-                        c2 == ContractType(c)
+                        contractIsSubtype(table, c1, c2, isThis) && ss1.subsetOf(ss2)
+                    case (StateType(c1, ss1, _), ContractReferenceType(c2, c2p, _)) =>
+                        contractIsSubtype(table, c1, c2, isThis)
                     case _ => false
                 }
-                if (!mainSubtype) Some(SubtypingError(t1, t2, isThis))
-                // TODO: make sure this is unnecessary: else if (!modifierSubtype) Some(OwnershipSubtypingError(t1, t2))
-                else None
-            case _ => Some(SubtypingError(t1, t2, isThis))
+
+            case (p: PrimitiveType, np: NonPrimitiveType) => np.contractName == "Top"
```

Comment:
Avoid hardcoding "Top" in various places; define an identifier for it somewhere (probably in ObsidianTypes).

Reply:
Actually I suggest calling this "Contract", analogous to Java's "Object".